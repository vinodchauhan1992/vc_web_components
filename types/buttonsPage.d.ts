import { VCToggleButtonProps } from './vcToggleButton';

export interface ButtonData {
  heading: string;
  btnsArr: VCToggleButtonProps[];
}
