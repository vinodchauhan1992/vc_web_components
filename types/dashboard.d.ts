export interface DashboardComponentListItemProps {
  label: string;
  route?: string;
}

export interface DashboardComponentListProps {
  label: string;
  items?: DashboardComponentListItemProps[];
}
