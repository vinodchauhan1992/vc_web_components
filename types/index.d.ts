import { ReactNode } from 'react';
import {
  AppBreadcrumbProps,
  AppBreadcrumbState,
  AppConfigProps,
  AppMenuItem,
  AppMenuItemProps,
  AppSubMenuProps,
  AppTopbarRef,
  Breadcrumb,
  BreadcrumbItem,
  ChatContextProps,
  LayoutConfig,
  LayoutContextProps,
  LayoutState,
  MailContextProps,
  MenuContextProps,
  MenuModel,
  MenuModelItem,
  MenuProps,
  NodeRef,
  Page,
  TaskContextProps,
} from './layout';

type ChildContainerProps = {
  children: ReactNode;
};

export type {
  AppBreadcrumbProps,
  AppBreadcrumbState,
  AppConfigProps,
  AppMailProps,
  AppMailReplyProps,
  AppMailSidebarItem,
  AppMenuItem,
  AppMenuItemProps,
  AppSubMenuProps,
  AppTopbarRef,
  Breadcrumb,
  Breadcrumb,
  BreadcrumbItem,
  ChartDataState,
  ChartOptionsState,
  ChatContextProps,
  ChildContainerProps,
  CustomEvent,
  Demo,
  LayoutConfig,
  LayoutContextProps,
  LayoutState,
  LayoutType,
  MailContextProps,
  MenuContextProps,
  MenuModel,
  MenuModelItem,
  MenuProps,
  NodeRef,
  Page,
  SortOrderType,
  TaskContextProps,
};
