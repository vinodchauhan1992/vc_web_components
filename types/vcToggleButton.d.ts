export type VCToggleButtonAppearanceProps = 'default' | 'subtle';

export type VCToggleButtonStateProps = 'default' | 'hover' | 'pressed';

export type VCToggleButtonSizeProps = 'xs' | 'sm' | 'md' | 'lg';

export type VCToggleButtonWidthProps = 'intrinsic' | 'full';

export interface VCToggleButtonProps {
  appearance?: VCToggleButtonAppearanceProps;
  state?: VCToggleButtonStateProps;
  size?: VCToggleButtonSizeProps;
  checked?: boolean;
  width?: VCToggleButtonWidthProps;
  disabled?: boolean;
  loading?: boolean;
  iconOnly?: boolean;
  leadingIcon?: boolean;
  trailingIcon?: boolean;
  label?: string;
}
