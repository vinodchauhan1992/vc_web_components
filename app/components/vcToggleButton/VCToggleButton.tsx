'use client';

import { VCToggleButtonProps } from '@/types/vcToggleButton';
import { useState } from 'react';

const VCToggleButton = (props: VCToggleButtonProps) => {
  const {
    appearance = 'default',
    size = 'md',
    checked = false,
    width = 'intrinsic',
    disabled = false,
    iconOnly = false,
    label = '',
    onClicked,
    trailingIcon,
    leadingIcon,
  } = props;

  const [btnChecked, setBtnChecked] = useState(checked);

  const getCommonBtnClass = (className: string) => {
    let prevClass = className;
    if (width === 'full' && size !== 'xs' && !iconOnly) {
      prevClass = `${prevClass} vc-toggle-full-width-btn`;
    }
    if (width === 'intrinsic') {
      prevClass = `${prevClass} vc-toggle-intrinsic-width-btn`;
    }
    if (iconOnly) {
      prevClass = `${prevClass} vc-toggle-icon-only-btn-common vc-toggle-${size}-icon-only-btn`;
    }
    if (!iconOnly) {
      prevClass = `${prevClass} vc-toggle-${size}-label-btn`;
    }
    return prevClass;
  };

  const getDisabledBtnClass = () => {
    let className = 'vc-toggle-btn-common';
    if (appearance === 'default') {
      className = `${className} disabled-vc-toggle-btn-common`;
    }
    if (appearance === 'subtle') {
      className = `${className} disabled-subtle-vc-toggle-btn-common`;
    }

    const newClass = getCommonBtnClass(className);
    console.log('newClass', newClass, width);
    return newClass;
  };

  const getDefaultAppearanceBtnClass = () => {
    let className = 'vc-toggle-btn-common';
    if (checked) {
      className = `${className} default-checked-vc-toggle-btn-common`;
    } else {
      className = `${className} default-unchecked-vc-toggle-btn-common`;
    }

    return getCommonBtnClass(className);
  };

  const getSubtleAppearanceBtnClass = () => {
    let className = 'vc-toggle-btn-common';
    if (checked) {
      className = `${className} subtle-checked-vc-toggle-btn-common`;
    } else {
      className = `${className} subtle-unchecked-vc-toggle-btn-common`;
    }

    return getCommonBtnClass(className);
  };

  const getBtnClassName = () => {
    if (disabled) {
      return getDisabledBtnClass();
    }
    if (appearance === 'default') {
      return getDefaultAppearanceBtnClass();
    }
    return getSubtleAppearanceBtnClass();
  };

  const renderLabel = () => {
    if ((trailingIcon || leadingIcon) && !iconOnly) {
      return (
        <div
          role="toggle_btn_label_icon_wrapper_div"
          className="vc-toggle-wrapper-trailing-leading-icon"
        >
          {leadingIcon ? (
            <div role="toggle_btn_leading_icon_div" className="vc-toggle-leading-icon-wrapper">
              <i
                role="toggle_btn_leading_icon"
                className={`pi pi-angle-down vc-toggle-${size}-trailing-leading-icon`}
              />
            </div>
          ) : null}
          <div role="toggle_btn_label_icon_label_div" className="vc-toggle-label-wrapper">
            {label}
          </div>
          {trailingIcon ? (
            <div role="toggle_btn_trailing_icon_div" className="vc-toggle-trailing-icon-wrapper">
              <i
                role="toggle_btn_trailing_icon"
                className={`pi pi-angle-down vc-toggle-${size}-trailing-leading-icon`}
              />
            </div>
          ) : null}
        </div>
      );
    }
    return (
      <div role="toggle_btn_label_div" className="vc-toggle-label-wrapper">
        {label}
      </div>
    );
  };

  const renderChildren = () => {
    return iconOnly ? <i className={`pi pi-image vc-toggle-${size}-icon`} /> : renderLabel();
  };

  const onClick = () => {
    if (!disabled) {
      if (onClicked) {
        onClicked({ checked: !checked });
      }
      setBtnChecked(!btnChecked);
    }
  };

  return (
    <div role="toggle_btn_wrapper_div">
      <div role="toggle_btn_container_div" className={getBtnClassName()} onClick={onClick}>
        {renderChildren()}
      </div>
    </div>
  );
};

export default VCToggleButton;
