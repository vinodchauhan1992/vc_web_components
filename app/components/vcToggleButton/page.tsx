'use client';

import { VCToggleButtonProps } from '@/types/vcToggleButton';

const VCToggleButton = (props: VCToggleButtonProps) => {
  const {
    appearance = 'default',
    state = 'default',
    size = 'md',
    checked = false,
    width = 'intrinsic',
    disabled = false,
    loading = false,
    iconOnly = false,
    leadingIcon = false,
    trailingIcon = false,
    label = '',
  } = props;

  const getDisabledBtnClass = () => {
    let className = 'vc-toggle-btn-common disabled-vc-toggle-btn-common';
    if (width === 'full' && size !== 'xs' && !iconOnly) {
      className = `${className} vc-toggle-full-width-btn`;
    }
    if (width === 'intrinsic') {
      className = `${className} vc-toggle-intrinsic-width-btn`;
    }
    if (iconOnly) {
      className = `${className} vc-toggle-icon-only-btn-common vc-toggle-${size}-icon-only-btn`;
    }
    if (!iconOnly) {
      className = `${className} vc-toggle-${size}-label-btn`;
    }

    return className;
  };

  const getBtnClassName = () => {
    if (disabled) {
      return getDisabledBtnClass();
    }
    return '';
  };

  const renderChildren = () => {
    return iconOnly ? <i className={`pi pi-image vc-toggle-${size}-icon`} /> : <div>{label}</div>;
  };

  return (
    <div>
      <div className={getBtnClassName()}>{renderChildren()}</div>
    </div>
  );
};

export default VCToggleButton;
