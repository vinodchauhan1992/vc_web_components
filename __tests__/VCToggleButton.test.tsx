import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import React from 'react';
import VCToggleButton from '../app/components/vcToggleButton/VCToggleButton';
import { VCToggleButtonProps } from '../types/vcToggleButton';

describe('VCToggleButton component tests', () => {
  const props: VCToggleButtonProps = {
    label: 'Toggle me',
    onClicked: jest.fn(),
  };

  const fetchClassName = (role = 'toggle_btn_container_div') => {
    const toggleBtnDiv = screen.getByRole(role);
    const className = toggleBtnDiv.className;
    return className;
  };

  it('renders component', () => {
    const { container } = render(<VCToggleButton {...props} />);

    expect(container).toMatchSnapshot();
  });

  describe('Default appearance toggle button tests', () => {
    const defaultAppearanceProps: VCToggleButtonProps = {
      ...props,
      appearance: 'default',
    };

    it('should render xs button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...defaultAppearanceProps,
        size: 'xs',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common default-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-xs-label-btn'
      );
    });

    it('should render sm button with icons only', () => {
      const updatedProps: VCToggleButtonProps = {
        ...defaultAppearanceProps,
        size: 'md',
        iconOnly: true,
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common default-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-icon-only-btn-common vc-toggle-md-icon-only-btn'
      );
    });

    it('should render md button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...defaultAppearanceProps,
        size: 'md',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common default-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-md-label-btn'
      );
    });

    it('should render lg button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...defaultAppearanceProps,
        size: 'lg',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common default-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-lg-label-btn'
      );
    });

    it('should render lg button with full width', () => {
      const updatedProps: VCToggleButtonProps = {
        ...defaultAppearanceProps,
        size: 'lg',
        width: 'full',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common default-unchecked-vc-toggle-btn-common vc-toggle-full-width-btn vc-toggle-lg-label-btn'
      );
    });

    it('should render md button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...defaultAppearanceProps,
        size: 'md',
        checked: true,
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common default-checked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-md-label-btn'
      );
    });
  });

  describe('Subtle appearance toggle button tests', () => {
    const subtleAppearanceProps: VCToggleButtonProps = {
      ...props,
      appearance: 'subtle',
    };

    it('should render xs button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...subtleAppearanceProps,
        size: 'xs',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common subtle-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-xs-label-btn'
      );
    });

    it('should render sm button with icons only', () => {
      const updatedProps: VCToggleButtonProps = {
        ...subtleAppearanceProps,
        size: 'md',
        iconOnly: true,
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common subtle-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-icon-only-btn-common vc-toggle-md-icon-only-btn'
      );
    });

    it('should render md button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...subtleAppearanceProps,
        size: 'md',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common subtle-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-md-label-btn'
      );
    });

    it('should render lg button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...subtleAppearanceProps,
        size: 'lg',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common subtle-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-lg-label-btn'
      );
    });

    it('should render lg button with full width', () => {
      const updatedProps: VCToggleButtonProps = {
        ...subtleAppearanceProps,
        size: 'lg',
        width: 'full',
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common subtle-unchecked-vc-toggle-btn-common vc-toggle-full-width-btn vc-toggle-lg-label-btn'
      );
    });

    it('should render checked md button', () => {
      const updatedProps: VCToggleButtonProps = {
        ...subtleAppearanceProps,
        size: 'md',
        checked: true,
      };

      render(<VCToggleButton {...updatedProps} />);
      expect(fetchClassName()).toBe(
        'vc-toggle-btn-common subtle-checked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-md-label-btn'
      );
    });
  });

  it('should render label', () => {
    const updatedProps: VCToggleButtonProps = {
      ...props,
      size: 'xs',
      appearance: 'default',
      checked: true,
    };

    const { container } = render(<VCToggleButton {...updatedProps} />);
    expect(container).toMatchSnapshot();

    const label = screen.getByText('Toggle me');
    expect(label).toBeInTheDocument();
  });

  it('should render toggle button with leading and trailing icon', () => {
    const updatedProps: VCToggleButtonProps = {
      ...props,
      label: 'Toggled on',
      size: 'md',
      appearance: 'default',
      checked: true,
      leadingIcon: true,
      trailingIcon: true,
    };

    const { container } = render(<VCToggleButton {...updatedProps} />);
    expect(container).toMatchSnapshot();

    const label = screen.getByText('Toggled on');
    expect(label).toBeInTheDocument();

    const leadingIconDivClassName = fetchClassName('toggle_btn_leading_icon_div');
    const leadingIconClassName = fetchClassName('toggle_btn_leading_icon');
    const trailingIconDivClassName = fetchClassName('toggle_btn_trailing_icon_div');
    const trailingIconClassName = fetchClassName('toggle_btn_leading_icon');
    const labelDivClassName = fetchClassName('toggle_btn_label_icon_label_div');

    expect(leadingIconDivClassName).toBe('vc-toggle-leading-icon-wrapper');
    expect(leadingIconClassName).toBe('pi pi-angle-down vc-toggle-md-trailing-leading-icon');
    expect(trailingIconDivClassName).toBe('vc-toggle-trailing-icon-wrapper');
    expect(trailingIconClassName).toBe('pi pi-angle-down vc-toggle-md-trailing-leading-icon');
    expect(labelDivClassName).toBe('vc-toggle-label-wrapper');
  });
});
