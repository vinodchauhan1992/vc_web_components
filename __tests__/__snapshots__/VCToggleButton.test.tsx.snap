// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCToggleButton component tests renders component 1`] = `
<div>
  <div
    role="toggle_btn_wrapper_div"
  >
    <div
      class="vc-toggle-btn-common default-unchecked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-md-label-btn"
      role="toggle_btn_container_div"
    >
      <div
        class="vc-toggle-label-wrapper"
        role="toggle_btn_label_div"
      >
        Toggle me
      </div>
    </div>
  </div>
</div>
`;

exports[`VCToggleButton component tests should render label 1`] = `
<div>
  <div
    role="toggle_btn_wrapper_div"
  >
    <div
      class="vc-toggle-btn-common default-checked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-xs-label-btn"
      role="toggle_btn_container_div"
    >
      <div
        class="vc-toggle-label-wrapper"
        role="toggle_btn_label_div"
      >
        Toggle me
      </div>
    </div>
  </div>
</div>
`;

exports[`VCToggleButton component tests should render toggle button with leading and trailing icon 1`] = `
<div>
  <div
    role="toggle_btn_wrapper_div"
  >
    <div
      class="vc-toggle-btn-common default-checked-vc-toggle-btn-common vc-toggle-intrinsic-width-btn vc-toggle-md-label-btn"
      role="toggle_btn_container_div"
    >
      <div
        class="vc-toggle-wrapper-trailing-leading-icon"
        role="toggle_btn_label_icon_wrapper_div"
      >
        <div
          class="vc-toggle-leading-icon-wrapper"
          role="toggle_btn_leading_icon_div"
        >
          <i
            class="pi pi-angle-down vc-toggle-md-trailing-leading-icon"
            role="toggle_btn_leading_icon"
          />
        </div>
        <div
          class="vc-toggle-label-wrapper"
          role="toggle_btn_label_icon_label_div"
        >
          Toggled on
        </div>
        <div
          class="vc-toggle-trailing-icon-wrapper"
          role="toggle_btn_trailing_icon_div"
        >
          <i
            class="pi pi-angle-down vc-toggle-md-trailing-leading-icon"
            role="toggle_btn_trailing_icon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
